vignettes/commands-and-events.Rmd

```diff
@@ -0,0 +1,88 @@
+---
+title: "Commands and events"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Commands and Events}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+
+editor:
+  markdown:
+    wrap: sentence
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+# Commands and Events
+
+The Chrome DevTools Protocol has two types of methods: *commands* and *events*.
+Methods like `Page$navigate()` and `DOM$querySelector()` are **commands**.
+That is, they tell the browser to do something; the browser does it, and then sends back some data.
+
+**Events** are quite different from commands.
+When, for example, you run `b$Page$loadEventFired()`, it does not send a message to the browser.
+Rather, this method tells the R process to wait until it receives a `Page.loadEventFired` message from the browser.
+
+Here is an example of how that event can be used.
+Note that these two lines of code must be run together, without any delay at all (this can be enforced by wrapping both lines of code in `{ .... }`).
+
+``` r
+library(chromote)
+b <- ChromoteSession$new()
+
+# Send a command to navigate to a page
+b$Page$navigate("https://www.r-project.org")
+#> $frameId
+#> [1] "0ADE3CFBAF764B0308ADE1ACCC33358B"
+#>
+#> $loaderId
+#> [1] "112AF4AC0C13FF4A95BED8173C3F4C7F"
+
+# Wait for the Page.loadEventFired event
+b$Page$loadEventFired()
+#> $timestamp
+#> [1] 680.7603
+```
+
+After running these two lines, the R process will be blocked.
+While it's blocked, the browser will load the page, and then send a message to the R process saying that the `Page.loadEventFired` event has occurred.
+The message looks something like this:
+
+``` json
+{"method":"Page.loadEventFired","params":{"timestamp":699232.345338}}
+```
+
+After the R process receives this message, the function returns the value, which looks like this:
+
+```         
+$timestamp
+[1] 699232.3
+```
+
+> **Note:** This sequence of commands, with `Page$navigate()` and then `Page$loadEventFired()` will not work 100% of the time.
+> See [Loading a Page Reliably](#loading-a-page-reliably) for more information.
```

Comment:
I think this link target needs to be fixed.

Reply:
Thanks for the reminder! I did a full text search and found another example, so I think I got them all now.